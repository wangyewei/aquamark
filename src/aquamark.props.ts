export type AquamarkProps = {

}

const defaultProps = (): AquamarkProps => ({})
export function defineProps(props: AquamarkProps = defaultProps()) {
  return props
}