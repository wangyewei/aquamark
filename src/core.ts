import type { AquamarkProps } from "./aquamark.props"

const CANVAS_ID = 'aquamark-root'
export function render(configs: AquamarkProps) {
  const CANVAS_EL = document.querySelector(`#${CANVAS_ID}`)
  if (CANVAS_EL) {
    document.body.removeChild(CANVAS_EL)
  }
  const canvas = document.createElement('canvas')
  draw(canvas, configs)
}

function draw(
  canvas: HTMLCanvasElement,
  props: AquamarkProps
) {
  const ctx = canvas.getContext('2d')
  if (!ctx) return
  const { content } = props
  // TODO: all of the follwing attributes should be configurable
  ctx.rotate(-20 * Math.PI / 180)
  ctx.font = '24px Microsoft JhengHei'
  ctx.fillStyle = '#dddddd'
  ctx.textAlign = 'left'
  // TODO: remove the const 350
  ctx.fillText(content, 350 / 3, 350 / 3)
  const dom = document.createElement('div')
  dom.id = CANVAS_ID
  dom.style.pointerEvents = 'none'
  dom.style.top = '70px'
  dom.style.left = '0px'
  dom.style.position = 'fixed'
  dom.style.zIndex = '100000'
  dom.style.width = document.documentElement.clientWidth - 100 + 'px'
  dom.style.height = document.documentElement.clientHeight - 100 + 'px'
  dom.style.background = 'url(' + canvas.toDataURL('image/png') + ') left top repeat'
  document.body.appendChild(dom)
  return CANVAS_ID
}