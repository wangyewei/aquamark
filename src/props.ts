export type AquamarkProps = {
  content: string
}

const defaultProps = (): AquamarkProps => ({
  content: 'Aquamark'
})
export const defineProps = (props: AquamarkProps = defaultProps()) => props