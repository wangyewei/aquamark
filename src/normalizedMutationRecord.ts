import { CANVAS_ID } from "./core"

export function normalizedMutationRecord(record: MutationRecord[]): boolean[] {
  const r: boolean[] = []

  for (const item of record) {
    if (item?.removedNodes && item.removedNodes.length) {
      for (const node of item.removedNodes) {
        if ((node as HTMLElement)?.id === CANVAS_ID) {
          r.push(true)
          break
        }
      }
    }
    const ob = item.target as HTMLElement
    if (ob?.id && ob.id === CANVAS_ID) {
      r.push(true)
      break
    }
  }

  return r
}