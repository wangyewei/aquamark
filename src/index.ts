import { defineProps, type AquamarkProps } from './aquamark.props'
import { render } from './core'

export default class Aquamark {
  #configs: AquamarkProps

  constructor(configs: AquamarkProps) {
    this.#configs = defineProps(configs)
  }

  init() {
    this.#renderMark(this.#configs)
  }

  #renderMark(configs: AquamarkProps) {
    render(configs)
  }
}


