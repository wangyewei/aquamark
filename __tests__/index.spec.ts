/**
 * @vitest-environment jsdom
 */
import { describe, expect, test, vi, beforeAll } from 'vitest'
import Aquamark from '../src';
import { CANVAS_ID, render } from '../src/core';
import { defineProps } from '../src/props';


function sleep(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

describe('Aquamark Basic Usage', () => {
  let aquamark: Aquamark;

  beforeAll(() => {
    aquamark = new Aquamark();
    aquamark.init()
  });

  test('Aquamark should be rendered', async () => {
    const el = document.getElementById(CANVAS_ID)
    expect(el?.outerHTML).toMatchSnapshot()
  });

  test('Render should not render multi marks', () => {
    const orgingalEl = document.querySelectorAll(`#${CANVAS_ID}`)
    expect(orgingalEl.length).toEqual(1)
    render(defineProps({}))
    const latestEl = document.querySelectorAll(`#${CANVAS_ID}`)
    expect(latestEl.length).toEqual(1)
  })

  test('Aquamark cannot be changed', async () => {
    const spy = vi.spyOn(aquamark, 'init')
    await sleep(100)
    const el = document.getElementById(CANVAS_ID)!
    expect(spy).toBeCalledTimes(0)
    el.style.opacity = '0'
    await sleep(100)
    const elNew = document.getElementById(CANVAS_ID)!
    expect(elNew.style.opacity).toBe('')
    expect(elNew.outerHTML).not.contain('oapcity')
    expect(spy).toBeCalledTimes(1)
    expect(elNew).toMatchSnapshot()
  })

  test('Auamark cannot be removed', async () => {
    const el = document.getElementById(CANVAS_ID)!
    document.body.removeChild(el)
    await sleep(100)
    const newEl = document.getElementById(CANVAS_ID)
    expect(newEl).toBeDefined()
    expect(newEl?.outerHTML).toMatchSnapshot()
  })
})