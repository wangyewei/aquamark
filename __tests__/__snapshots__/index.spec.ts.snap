// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Aquamark Basic Usage > Aquamark cannot be changed 1`] = `
<div
  id="aquamark-root"
  style="pointer-events: none; top: 0px; left: 0px; position: fixed; z-index: 100; width: 0px; height: 0px; background: url(data:image/png;base64,00) repeat top;"
/>
`;

exports[`Aquamark Basic Usage > Aquamark should be rendered 1`] = `"<div id="aquamark-root" style="pointer-events: none; top: 0px; left: 0px; position: fixed; z-index: 100; width: 0px; height: 0px; background: url(data:image/png;base64,00) repeat top;"></div>"`;

exports[`Aquamark Basic Usage > Auamark cannot be removed 1`] = `"<div id="aquamark-root" style="pointer-events: none; top: 0px; left: 0px; position: fixed; z-index: 100; width: 0px; height: 0px; background: url(data:image/png;base64,00) repeat top;"></div>"`;
